fdb-extensions/src/test/java/com/apple/foundationdb/async/RTreeModificationTest.java

```diff
@@ -0,0 +1,352 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.async;
+
+import com.apple.foundationdb.Database;
+import com.apple.foundationdb.FDB;
+import com.apple.foundationdb.FDBTestBase;
+import com.apple.foundationdb.KeyValue;
+import com.apple.foundationdb.NetworkOptions;
+import com.apple.foundationdb.Range;
+import com.apple.foundationdb.directory.DirectoryLayer;
+import com.apple.foundationdb.directory.DirectorySubspace;
+import com.apple.foundationdb.directory.PathUtil;
+import com.apple.foundationdb.subspace.Subspace;
+import com.apple.foundationdb.tuple.Tuple;
+import com.apple.test.Tags;
+import com.google.common.base.Verify;
+import com.google.common.collect.ImmutableList;
+import org.davidmoten.hilbert.HilbertCurve;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Tag;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+import org.junit.jupiter.params.provider.ValueSource;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+import java.math.BigInteger;
+import java.util.List;
+import java.util.Random;
+import java.util.concurrent.ForkJoinPool;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.stream.Stream;
+
+/**
+ * Tests for {@link RTree}.
+ */
+@Tag(Tags.RequiresFDB)
+public class RTreeModificationTest extends FDBTestBase {
+    private static final Logger logger = LoggerFactory.getLogger(RTreeModificationTest.class);
+
+    private static final int NUM_TEST_RUNS = 10;
+    private static final int NUM_SAMPLES = 10_000;
+
+    private Database db;
+    private DirectorySubspace rtSubspace;
+
+    private static final boolean TRACE = false;
+
+    @BeforeEach
+    public void setUpDb() throws Exception {
+        FDB fdb = FDB.instance();
+        if (TRACE) {
+            NetworkOptions options = fdb.options();
+            options.setTraceEnable("/tmp");
+            options.setTraceLogGroup("RTreeTest");
+        }
+        db = fdb.open();
+        rtSubspace = DirectoryLayer.getDefault().createOrOpen(db, PathUtil.from(RTree.class.getSimpleName())).get();
+        db.run(tr -> {
+            tr.clear(Range.startsWith(rtSubspace.getKey()));
+            return null;
+        });
+    }
+
+    @AfterEach
+    public void closeDb() {
+        db.close();
+    }
+
+    @ParameterizedTest
+    @ValueSource(ints = {10, 100, 1000, 10000})
+    public void testAllDeleted(final int numSamples) {
```

Comment:
One other test that might make sense would be a random test that validated the integrity of the data structure. Something like:

1. The membership of the R-tree is as expected
2. All (non-root) pages have between `minM` and `maxM` values
3. Relationships between parent MBRs and child data are correctly handled

I will say that I reviewed some of the logic detailing specific rebalancing/splitting/merging cases less than some others, and having some stress tests of that would be useful in gaining confidence

Reply:
I added stress tests for 1. without and with deletes. 2. is checked when a node is fetched; 3 is also checked during and to some extent we also check that we don't get any data corpses lying around.

Reply:
I also added a `validate()` method to `RTree` and test cases.